MantidQt/MantidWidgets/src/SlitCalculator.cpp

```diff
@@ -1,48 +1,122 @@
+  Q_UNUSED(parent);
+  ui.setupUi(this);
+  if (currentInstrumentName == "") {
+    // set up the initial instrument if there is not one associated
+    // with slit calculator
+    currentInstrumentName = "INTER";
+    setInstrument(currentInstrumentName);
+  }
+  on_recalculate_triggered();
+}
+void SlitCalculator::processInstrumentHasBeenChanged() {
+  // used in refl main window to indicate that slitCalculator fields
+  // need to update because another instrument has been selected.
+  on_recalculate_triggered();
+}
+SlitCalculator::~SlitCalculator() {}
+void SlitCalculator::setInstrument(std::string instrumentName) {
+  // we want to get the most up-to-date definition, so we use the current
+  // date/time
+  auto date = Mantid::Kernel::DateAndTime::getCurrentTime().toISO8601String();
+  // find the full path to the definition file
+  auto filename =
+      Mantid::API::ExperimentInfo::getInstrumentFilename(instrumentName, date);
+  // parse the XML that we have found for the definition
+  Mantid::Geometry::InstrumentDefinitionParser parser =
+      Mantid::Geometry::InstrumentDefinitionParser(
+          filename, instrumentName,
+          Mantid::Kernel::Strings::loadFile(filename));
+  // retrieving the mangled name of the instrument
+  std::string instrumentNameMangled = parser.getMangledName();
+  // See if we have a definition already in the InstrumentDataService
+  if (Mantid::API::InstrumentDataService::Instance().doesExist(
+          instrumentNameMangled)) {
+    // If it does, set the associated instrument to the one we have found.
+    this->instrument = Mantid::API::InstrumentDataService::Instance().retrieve(
+        instrumentNameMangled);
+  } else {
+    // We set the instrument from XML that we have found.
+    Mantid::API::Progress *prog = new Mantid::API::Progress();
+    this->instrument = parser.parseXML(prog);
+    delete prog;
+  }
+  setupSlitCalculatorWithInstrumentValues(instrument);
+}
 
-    void SlitCalculator::on_recalculate_triggered()
-    {
-      //Gather input
-      const double s1s2 = ui.spinSlit1Slit2->value();
-      const double s2sa = ui.spinSlit2Sample->value();
-      const double res = ui.spinResolution->value();
-      const double footprint = ui.spinFootprint->value();
-      const double angle = ui.spinAngle->value();
+void SlitCalculator::setupSlitCalculatorWithInstrumentValues(
+    Mantid::Geometry::Instrument_const_sptr instrument) {
+  // fetch the components that we need for values from IDF
+  auto slit1Component = instrument->getComponentByName("slit1");
+  auto slit2Component = instrument->getComponentByName("slit2");
+  auto sampleComponent = instrument->getComponentByName("some-surface-holder");
+  // check that they have been fetched from the IDF
+  if (slit1Component.get() != NULL && slit2Component.get() != NULL &&
+      sampleComponent.get() != NULL) {
+    // convert from meters to millimeters
+    const double s1s2 = 1e3 * (slit1Component->getDistance(*slit2Component));
```

Comment:
Good. My only suggestion would have been to have a const static double called something like MiliMetersInMeter and to use that instead of the 1e3 temp.